### PR TITLE
Implementing Gather operator

### LINKED ISSUE #343: Support gather operation
A new op should be created to support https://pytorch.org/docs/stable/generated/torch.gather.html.

### LINKED ISSUE #596: Kernel refactor for `Gather`

### PR DESCRIPTION
**Pull request by jiazhihao**
**Description of changes:**

This PR provides an initial implementation for Gather (https://pytorch.org/docs/stable/generated/torch.gather.html). 

**Related Issues:**

Linked Issues:
- Issue #

Issues closed by this PR:
- Closes #596 
- Closes #343 

**Before merging:**

- [ ] Did you update the [flexflow-third-party](https://github.com/flexflow/flexflow-third-party) repo, if modifying any of the Cmake files, the build configs, or the submodules?

### CODE REVIEW

**Comment by lockshaw** [comment]
@soumyac1999

**Comment by jiazhihao** [comment]
@soumyac1999 I haven't tested the forward/backward kernel implementations for gather. Can you provide a tiny model using gather for testing purpose?

**Comment by gabrieleoliaro** [comment]
Should we rebase this on the master branch and remove the commits that were already merged?

**Comment by soumyac1999** [comment]
> @soumyac1999 I haven't tested the forward/backward kernel implementations for gather. Can you provide a tiny model using gather for testing purpose?

The following is roughly how I would use gather:

```python
import torch

neighbors = [[0, 5, 3], [3, 7, 9]]
flat_neighbors = [0, 5, 3, 3, 7, 9]

x = torch.rand(32, 16, 3)   # batch size = 32, graph with 16 nodes, each with 3 dimensional embedding
x0 = torch.nn.Linear(3, 5)(x)  # shape (32, 16, 5)

idxs = torch.tensor(flat_neighbors)
idxs = idxs.reshape((1, -1, 1)).repeat(32, 1, 5)

x1 = torch.gather(x0, 1, idxs)  # shape (32, 6, 5)
x1 = x1.reshape(-1, 2, 3, 5).mean(2)  # shape (32, 2, 5)

```

**Comment by jiazhihao** [comment]
@soumyac1999 Thanks! This gives me an example to test the implementation.

**Comment by jiazhihao** [comment]
@soumyac1999 PyTorch's gather require that `input` and `index` to have the same number of dimensions (https://pytorch.org/docs/stable/generated/torch.gather.html), and we are currently aligning our APIs with PyTorch's. You will need to update your program to satisfy the dimensionality constraints.

**Comment by soumyac1999** [comment]
Yes, I think the inputs and index have the same number of dimensions in the example (I tested in using pytorch version 1.12.1 and it seems to work).
```
>>> idxs.shape
torch.Size([32, 6, 5])
>>> x0.shape
torch.Size([32, 16, 5])
```

File: examples/python/native/test_gather.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  y = np.random.rand(bs, 6, 5).astype('float32')
+  label_tensor.attach_numpy_array(ffmodel, ffconfig, y)
+
+  for _ in range(100):
+    ffmodel.forward()
+    ffmodel.backward()
+    ffmodel.update()
+
+if __name__ == "__main__":
+  print("test reduce")
```
Comment (gabrieleoliaro) [inline examples/python/native/test_gather.py:34]: why does this line print "test reduce"? Should it be "test gather"? If not, I'd maybe add a comment to explain why

File: examples/python/native/test_gather.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+from flexflow.core import *
```
Comment (gabrieleoliaro) [inline examples/python/native/test_gather.py:1]: Very small comment, but should we rename this file to avoid using the word "test" if it's just an example? If, on the other hand, this is a test, should we move it to the `tests` folder to avoid crowding the examples folder too much?

**Review by gabrieleoliaro — CHANGES_REQUESTED** [review]
Overall LGTM! Just left a couple of comments. To test correctness, we should probably add the operator to the alignment tests, but we first need to fix the alignment testing framework as a whole, so we can do it in a separate PR. On the other hand, would it be useful to add the gather example/test to the `multi_gpu_tests.sh`?

**Comment by lockshaw** [comment]
> On the other hand, would it be useful to add the gather example/test to the multi_gpu_tests.sh?

I'm in favor of adding the example to `multi_gpu_tests.sh` for now, we can also remove it once pytorch alignment is up and running

File: src/ops/gather.cpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                            coord_t stride,
+                            coord_t dim_size,
+                            hipStream_t stream) {
+  assert(input_ptr != nullptr);
+  assert(index_ptr != nullptr);
+  assert(output_ptr != nullptr);
+  assert(false && "hip kernel to be implemented");
+}
+
+void Gather::backward_kernel_wrapper(GatherMeta const *m,
```
Comment (lockshaw) [inline src/ops/gather.cpp:76]: This seems to be not reliant at all on CUDA/HIP-specific behavior, so it should be moved into `gather.cc`

**Review by lockshaw — CHANGES_REQUESTED** [review]

File: examples/python/native/test_gather.py (hunk 1 above)
Comment (jiazhihao) [reply examples/python/native/test_gather.py]: Fixed.

File: examples/python/native/test_gather.py (hunk 2 above)
Comment (jiazhihao) [reply examples/python/native/test_gather.py]: Fixed.

File: examples/python/native/test_gather.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  ffmodel.optimizer = ffoptimizer
+  ffmodel.compile(loss_type=LossType.LOSS_MEAN_SQUARED_ERROR_AVG_REDUCE, metrics=[MetricsType.METRICS_MEAN_SQUARED_ERROR])
+  ffmodel.init_layers()
+  input.attach_numpy_array(ffmodel, ffconfig, x)
+  index.attach_numpy_array(ffmodel, ffconfig, neighbors)
+  label_tensor = ffmodel.label_tensor
+  y = np.random.rand(bs, 6, 5).astype('float32')
+  label_tensor.attach_numpy_array(ffmodel, ffconfig, y)
+
+  for _ in range(100):
```
[the remark this answers is not in the document]
Comment (jiazhihao) [reply examples/python/native/test_gather.py]: The program train a model with fixed input; I just let it run for a number of iterations and make sure we can observe a decreasing loss, meaning the model overfit to the data.

[the remark this answers is not in the document]
Comment (jiazhihao) [reply examples/python/native/test_gather.py]: This corresponds to 100 forward/backward iterations and not 100 epoches.

File: include/flexflow/ops/gather.h [hunk 5]
```diff
@@ -0,0 +1,96 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (jiazhihao) [reply include/flexflow/ops/gather.h:1]: Fixed.

File: include/flexflow/ops/gather.h [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  Gather(FFModel &model,
+         Params const &params,
+         Input const &input,
+         char const *name = nullptr);
+  Gather(FFModel &model,
+         const ParallelTensor input,
+         const ParallelTensor index,
+         int legion_dim,
+         char const *name = nullptr);
+  void init(FFModel const &) override;
```
[the remark this answers is not in the document]
Comment (jiazhihao) [reply include/flexflow/ops/gather.h:31]: I am planning to make it a separate PR. The current plan is to merge this to the master branch so @soumyac1999 can start importing his model.

File: python/flexflow/type.py [hunk 7]
```diff
@@ -103,6 +103,7 @@ class OpType(Enum):
   UNSQUEEZE = 2103
   TYPE_AS = 2104
   VIEW = 2105
+  GATHER = 2106
```
[the remark this answers is not in the document]
Comment (jiazhihao) [reply python/flexflow/type.py:106]: @eddy16112 may know more about this, but my understanding is that the C++ values don't have to be identical to the Python values.

File: src/ops/gather.cc [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+Gather::Gather(FFModel &model,
+               GatherParams const &params,
+               std::pair<ParallelTensor, ParallelTensor> const &inputs,
+               char const *name)
+    : Gather(model, inputs.first, inputs.second, params.legion_dim, name) {}
+
+Gather::Gather(FFModel &model,
+               const ParallelTensor input,
+               const ParallelTensor index,
+               int _legion_dim,
```
[the remark this answers is not in the document]
Comment (jiazhihao) [reply src/ops/gather.cc:122]: SGTM.

File: src/ops/gather.cpp [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+                           output.get_float_ptr(),
+                           output.domain.get_volume(),
+                           stride,
+                           dim_size,
+                           stream);
+  }
+}
+
+template <typename TI>
+void Gather::forward_kernel(float const *input_ptr,
```
[the remark this answers is not in the document]
Comment (jiazhihao) [reply src/ops/gather.cpp:63]: See if you like my fix for this.

**Comment by jiazhihao** [comment]
@lockshaw @gabrieleoliaro A kind reminder on this PR.

**Comment by gabrieleoliaro** [comment]
> @lockshaw @gabrieleoliaro A kind reminder on this PR.

The changes look good to me. @jiazhihao Could you just add the demo_gather.py example to `tests/multi_gpu_tests.sh`? I was going to do it myself, but I don't think I can push to your fork

**Comment by jiazhihao** [comment]
I think we should not include `demo_gather.py` as it tests Gather using `attach_tensor` instead of data loader. Therefore, it is not a training task but simply forward/backward runs.

**Review by gabrieleoliaro — APPROVED** [review]

File: include/flexflow/ops/gather.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (jiazhihao) [reply include/flexflow/ops/gather.h:32]: ping @lockshaw

File: src/ops/gather.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (lockshaw) [reply src/ops/gather.cpp:63]: I think this is fine for now, but in the long term we should figure out how best to handle the different support for different backends in a more visible way. Let's merge this and continue the discussion in #595.

**Comment by lockshaw** [comment]
>

**Comment by lockshaw** [comment]
> I think we should not include `demo_gather.py` as it tests Gather using `attach_tensor` instead of data loader. Therefore, it is not a training task but simply forward/backward runs.

Can you clarify why not? Considering the current scarcity of tests, I'm kinda on board to add any and all tests and refine them later, but maybe I misunderstand the issue here

File: src/ops/gather.cc [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+
+Op *Gather::materialize(FFModel &ff,
+                        ParallelTensor inputs[],
+                        int num_inputs) const {
+  GatherParams params = get_params();
+  return new Gather(ff, params, {inputs[0], inputs[1]}, this->name);
+}
+
+void Gather::init(FFModel const &ff) {
+  assert(check_output_input_weight_same_parallel_is());
```
[the remark this answers is not in the document]
Comment (lockshaw) [reply src/ops/gather.cc:179]: Moved to #583.

File: include/flexflow/ops/gather.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (lockshaw) [reply include/flexflow/ops/gather.h:32]: Moved to #596 and #303

File: python/flexflow/type.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (lockshaw) [reply python/flexflow/type.py:107]: Tracking in #597

File: examples/python/native/test_gather.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lockshaw) [reply examples/python/native/test_gather.py]: > The program train a model with fixed input; I just let it run for a number of iterations and make sure we can observe a decreasing loss, meaning the model overfit to the data.

Is this checked automatically (so it can be validated as part of CI)?

File: src/ops/gather.cc [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+                         int num_inputs) {
+  assert(num_inputs == 2);
+  int legion_dim;
+  dez.deserialize(legion_dim);
+  GatherParams params;
+  params.legion_dim = legion_dim;
+  return ff.get_or_create_node<Gather>({inputs[0], inputs[1]}, params);
+}
+
+Op *Gather::materialize(FFModel &ff,
```
[the remark this answers is not in the document]
Comment (lockshaw) [reply src/ops/gather.cc:171]: Why not just use a constructor?

File: src/runtime/hip_helper.cpp [hunk 12]
```diff
@@ -325,6 +325,12 @@ hipblasDatatype_t ff_to_cuda_datatype(DataType type) {
   }
   return HIPBLAS_R_32F;
 }
+
+void handle_unimplemented_hip_kernel() {
```
[the remark this answers is not in the document]
Comment (lockshaw) [reply src/runtime/hip_helper.cpp:329]: Also, https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#nl25-dont-use-void-as-an-argument-type for the header (again unless @eric-zheng or @gabrieleoliaro or anyone else objects)

**Review by lockshaw — CHANGES_REQUESTED** [review]

**Comment by lockshaw** [comment]
@jiazhihao Did another pass (left comments on code, unresolved conversations I still have questions on). Lmk when this is ready for another pass :slightly_smiling_face:. 

Also, while we're at it--could you add some brief doxygen docs explaining the operator class (don't have to do all of the internal functions) and linking to whatever pytorch/tf/keras/etc. docs make sense?

File: src/ops/gather.cc [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+  return gather->outputs[0];
+}
+
+Op *Gather::create_operator_from_layer(
+    FFModel &model,
+    Layer const *layer,
+    std::vector<ParallelTensor> const &inputs) {
+  long long value;
+  layer->get_int_property("legion_dim", value);
+  int legion_dim = value;
```
[the remark this answers is not in the document]
Comment (jiazhihao) [reply src/ops/gather.cc:112]: All integers are stored using type long long in `add_int_property` and `get_int_property`. Thus we narrow it back to type int (see L98 for adding this int property).

File: src/ops/gather.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (jiazhihao) [reply src/ops/gather.cc:174]: `materialize` is a virtual function of class `Op`, which is used for instantiating operators when constructing the final PCG (https://github.com/flexflow/FlexFlow/blob/master/src/runtime/substitution.cc#L3264)

File: include/flexflow/ops/gather_params.h [hunk 14]
```diff
@@ -0,0 +1,22 @@
+#pragma once
```
Comment (lockshaw) [inline include/flexflow/ops/gather_params.h:1]: `#ifndef` for now

**Review by lockshaw — APPROVED** [review]
LGTM, just fix the one `#pragma -> #ifndef` and then let's merge :slightly_smiling_face:

File: include/flexflow/ops/gather_params.h (hunk 14 above)
Comment (jiazhihao) [reply include/flexflow/ops/gather_params.h]: done
